lib/sidekiq/history/middleware.rb

```diff
@@ -10,56 +10,58 @@ def call(worker, msg, queue, &block)
       private
 
       def call_with_sidekiq_history(worker, msg, queue)
-        worker_status = new_status
+        worker_status = { last_job_status: 'passed'.freeze }
+
         start = Time.now.utc
 
         yield
       rescue StandardError => e
-        worker_status[:failed] = 1
-        worker_status[:passed] = 0
         worker_status[:last_job_status] = 'failed'.freeze
 
         raise e
       ensure
-        worker_status[:runtime] ||= [elapsed(start)]
+        worker_status[:time] = elapsed(start)
         worker_status[:last_runtime] = Time.now.utc
-        worker_status[:last_job_status] ||= 'passed'.freeze
         worker_status[:class] = msg['wrapped'.freeze] || worker.class.to_s
 
         save_entry_for_worker worker_status
       end
 
-      def new_status
-        {
-          failed: 0,
-          passed: 1
-        }
-      end
-
       def save_entry_for_worker(worker_status)
         status = worker_status.dup
-        worker = status.delete :class
+        worker_key = "#{Time.now.utc.to_date}:#{status.delete :class}"
 
         Sidekiq.redis do |redis|
-          history = "sidekiq:history:#{Time.now.utc.to_date}"
+          redis.watch(REDIS_HASH) do
+            min_time = hget_time(redis, "#{worker_key}:min_time", status)
+            max_time = hget_time(redis, "#{worker_key}:max_time", status)
+            average_time = hget_time(redis, "#{worker_key}:average_time", status)
 
-          redis.watch(history) do
-            value = redis.get history
+            statistics = [
+              "#{worker_key}:last_job_status", status[:last_job_status],
+              "#{worker_key}:average_time", (average_time + status[:time]) / 2,
```

Comment:
Instead of calculating all these time values on every single job execution, I would keep a list of the last N execution times.  That allows you to perform standard deviation, which is much more useful that average:

```ruby
redis.lpush "#{worker_key}:timings", status[:time]
redis.ltrim "#{worker_key}:timings", 0, 100
```